src/rdkafka_cgrp.c

```diff
@@ -3124,8 +3999,264 @@ rd_kafka_cgrp_max_poll_interval_check_tmr_cb (rd_kafka_timers_t *rkts,
+
+        /* TODO: Something that isn't O(N*M). */
+        for (i=0; i<unsubscribing->cnt; i++) {
+                for (j=0; j<rkcg->rkcg_assignment->cnt; j++) {
+                        rd_kafka_topic_partition_t *par =
+                                &rkcg->rkcg_assignment->elems[j];
+                        if (!strcmp(par->topic,
+                                    unsubscribing->elems[i].topic)) {
+                                rd_kafka_topic_partition_list_add(
+                                                        revoking,
+                                                        par->topic,
+                                                        par->partition);
+                                continue;
+                        }
+                }
+        }
+
+        if (revoking->cnt == 0) {
+                rd_kafka_topic_partition_list_destroy(revoking);
+                revoking = NULL;
+        }
+
+        return revoking;
+}
+// TODO: unit testing.
+
+
+/**
+ * @brief Handle a new subscription that is replacing an existing subscription
+ *        in the COOPERATIVE rebalance case.
+ */
+static rd_kafka_resp_err_t
+rd_kafka_cgrp_change_subscribed (rd_kafka_cgrp_t *rkcg,
+                                 rd_kafka_topic_partition_list_t *rktparlist) {
+        rd_kafka_topic_partition_list_t *unsubscribing;
+        rd_kafka_topic_partition_list_t *revoking;
+        rd_list_t *tinfos;
+        rd_kafka_topic_partition_list_t *errored;
+
+        // MH: true but may be suprfluous.
+        rkcg->rkcg_flags &= ~RD_KAFKA_CGRP_F_WILDCARD_SUBSCRIPTION;
+
+        if (rd_kafka_topic_partition_list_regex_cnt(rktparlist) > 0)
+                rkcg->rkcg_flags |= RD_KAFKA_CGRP_F_WILDCARD_SUBSCRIPTION;
+
+        /* determine the topics from rkcg_subscribed_topics that don't match
+         * an entry in rktparlist */
+        unsubscribing = rd_kafka_cgrp_calculate_unsubscribing_topics(
+                                        rkcg, rktparlist);
+
+        revoking = rd_kafka_cgrp_calculate_subscribe_revoking_partitions(
+                                        rkcg, unsubscribing);
+
+        rd_kafka_dbg(rkcg->rkcg_rk, CGRP|RD_KAFKA_DBG_CONSUMER, "SUBSCRIBE",
+                     "Group \"%.*s\": changing subscription from %d to %d "
+                     "topic(s)%s (join state %s)",
+                     RD_KAFKAP_STR_PR(rkcg->rkcg_group_id),
+                     rkcg->rkcg_subscription->cnt, rktparlist->cnt,
+                     !unsubscribing ? " with no topics removed" :
+                     (!revoking ? " without revoking any partitions" : ""),
+                     rd_kafka_cgrp_join_state_names[rkcg->rkcg_join_state]);
+
+        rd_kafka_topic_partition_list_destroy(rkcg->rkcg_subscription);
+        rkcg->rkcg_subscription = rktparlist;
+
+        /* Create a list of the topics in metadata that matches the new
+         * subscription */
+        tinfos = rd_list_new(rkcg->rkcg_subscription->cnt,
```

Comment:
fwiw the stuff here is not correct (if metadata needed). have this working locally now, kind of hacked in.